src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt

```diff
@@ -74,9 +77,106 @@ internal class RecordFetcherTest : AbstractVertxTest() {
             }
         }
 
-        val sut = RecordFetcher(kinesisClientMock, 1, "some-stream", ShardId("1")) { "" }
+        val sut = createFetcher(kinesisClientMock)
+
+        sut.fetchNextRecords(queryPosition)
+        val response = sut.getNextRecords()
+        response.nextShardIterator().shouldBe(ITERATOR)
+    }
+
+    @Test
+    internal fun getNextRecords_should_be_delayed(testContext: VertxTestContext) = asyncTest(testContext) {
+        val responseMock: GetRecordsResponse = mock {
+            on { nextShardIterator() } doReturn ITERATOR
+        }
+
+        val kinesisClientMock = mock<KinesisAsyncClient> {
+            on { getRecords(any<Consumer<GetRecordsRequest.Builder>>()) } doReturn CompletableFuture.completedFuture(
+                responseMock
+            )
+        }
+
+        val delay = 100L
 
-        val response = sut.fetchNextRecords(queryPosition)
-        response.nextShardIterator().shouldBe(nextIterator)
+        val sut = createFetcher(kinesisClientMock, delay)
+
+        val start = System.currentTimeMillis()
+
+        sut.fetchNextRecords(queryPosition)
+        sut.getNextRecords()
+
+        val end = System.currentTimeMillis()
+        val diff = end - start
+        diff.shouldBeGreaterThanOrEqual(delay)
     }
+
+    @Test
+    internal fun exception_case_restart_fetching(testContext: VertxTestContext) = asyncTest(testContext) {
```

Comment:
even after reading this test I still can't understand the underlying logic. I think more a docsting or possibly refactoring is needed.